cadence/freshmint-encoding/FreshmintEncoding.cdc

```diff
@@ -0,0 +1,105 @@
+/// FreshmintEncoding is a set of utilities for encoding Cadence values as byte arrays.
+///
+/// Variable-legnth values (String, Int, UInt) include a fixed-size length prefix to
+/// prevent distinct sets of values from encoding to the same byte sequence when concatenated.
+///
+/// For example, for a structure with two fields:
+///
+///  - foo: String
+///  - bar: String
+///
+/// Without a length prefix, these instances would produce the same encoding:
+///
+///  - Instance A: { foo: "foo", bar: "bar" }
+///  - Instance b: { foo: "foob", bar: "ar" }
+///
+/// when using an encoding function like this:
+/// 
+/// let encoding = encodeString(foo).concat(encodeString(bar))
+///
+pub contract FreshmintEncoding {
+
+    pub fun encodeUInt(_ value: UInt): [UInt8] {
+        let bytes = value.toBigEndianBytes()
+        let length = UInt16(bytes.length)
+
+        // Prefix byte encoding with a fixed-size encoding of its length
+        return length.toBigEndianBytes().concat(bytes)
+    }
+
+    pub fun encodeUInt8(_ value: UInt8): [UInt8] {
+        return value.toBigEndianBytes()
+    }
+
+    pub fun encodeUInt16(_ value: UInt16): [UInt8] {
+        return value.toBigEndianBytes()
```

Comment:
The validity of the encoding here depends on how `toBigEndianBytes` is implemented by Cadence, in particular, whether it always returns an output of the same size. 

For instance let's consider this struct `{ UInt16 a, UInt16 b}`.  Then the encodings of `{0x0102, 0x03}` and `{0x01, 0x0203}` may be equal. 
If `toBigEndianBytes` always returns 2 bytes for `UInt16` then it would be great to add tests to confirm it.

Reply:
The doc for `toBigEndianBytes` seems thin here : https://developers.flow.com/cadence/language/values-and-types

Reply:
All of these values product a fixed-length encoding left-padded with zeros (e.g. UInt16 is always 2 bytes, UInt32 is always 4 bytes).

```cdc
// Fixed-length encoding
Int8
Int16
Int32
Int64
UInt8
UInt16
UInt32
UInt64
Fix64
UFix64

// Variable-length encoding
Int
UInt
```

You can see test cases here, which are copied from the Cadence repository: https://github.com/packagelabs/freshmint/blob/alpha/packages/core/cadence/values.test.ts

I'll add tests that confirm that we get the same result from this smart contract.

Reply:
Nice!